File: java-library/src/main/java/com/gblninja/parse/ParseTag.java
package com.gblninja.parse;

import com.gblninja.results.ParseTagResult;
import com.gblninja.tag.TagHeader;
import com.gblninja.utils.ByteUtils;

public class ParseTag {
    private static final int TAG_ID_SIZE = 4;
    private static final int TAG_LENGTH_SIZE = 4;

    public static ParseTagResult parseTag(byte[] byteArray, int offset) {
        if (offset < 0 || offset + TAG_ID_SIZE + TAG_LENGTH_SIZE > byteArray.length) {
            return new ParseTagResult.Fatal("Invalid offset: " + offset);
        }

        long tagId = Integer.toUnsignedLong(ByteUtils.getIntFromBytes(byteArray, offset, TAG_ID_SIZE).getInt());
        int tagLength = ByteUtils.getIntFromBytes(byteArray, offset + TAG_ID_SIZE, TAG_LENGTH_SIZE).getInt();

        if (offset + TAG_ID_SIZE + TAG_LENGTH_SIZE + tagLength > byteArray.length) {
            return new ParseTagResult.Fatal("Invalid tag length: " + tagLength);
        }

        byte[] tagData = new byte[tagLength];
        System.arraycopy(byteArray, offset + TAG_ID_SIZE + TAG_LENGTH_SIZE, tagData, 0, tagLength);

        TagHeader tagHeader = new TagHeader(tagId, Integer.toUnsignedLong(tagLength));

        return new ParseTagResult.Success(tagHeader, tagData);
    }
}

File: gbl/src/main/java/com/onix/gbl/parser/parse/ParseTag.kt
package parser.data.parse

import parser.data.TAG_ID_SIZE
import parser.data.TAG_LENGTH_SIZE
import parser.data.results.ParseTagResult
import parser.data.tag.Tag
import parser.data.tag.TagHeader
import parser.data.type.GblType
import parser.data.utils.getFromBytes

fun parseTag(
    byteArray: ByteArray,
    offset: Int = 0,
): ParseTagResult {
    if (offset < 0 || offset + TAG_ID_SIZE + TAG_LENGTH_SIZE > byteArray.size) {
        return ParseTagResult.Fatal
    }

    val tagId = getFromBytes(byteArray, offset = offset, length = TAG_ID_SIZE).int

    val tagLength = getFromBytes(byteArray, offset = offset + TAG_ID_SIZE, length = TAG_LENGTH_SIZE).int

    if (offset + TAG_ID_SIZE + TAG_LENGTH_SIZE + tagLength > byteArray.size) {
        return ParseTagResult.Fatal
    }

    val tagData = byteArray.copyOfRange(
        offset + TAG_ID_SIZE + TAG_LENGTH_SIZE,
        offset + TAG_ID_SIZE + TAG_LENGTH_SIZE + tagLength
    )

    val tagHeader = TagHeader(
        id = tagId.toUInt(),
        length = tagLength.toUInt()
    )

    return ParseTagResult.Success(
        tag = Tag(
            tagHeader = tagHeader,
            tagType = GblType.TAG,
            tagData = tagData,
        )
    )
}
